src/modifier/default.rs

```diff
@@ -0,0 +1,537 @@
+    match (map.get(&Atom::Escape('A', 1)).unwrap(), map.get(&Atom::Escape('A', 2)).unwrap()) {
+        (Expression::Atom(Atom::Variable(a)), Expression::Atom(Atom::Numeric(b))) => (Expression::Atom(Atom::Numeric(*b)) * Expression::Atom(Atom::Variable(*a)), true),
+        (Expression::Atom(Atom::Variable(a)), Expression::Atom(Atom::Variable(b))) => (Expression::Atom(Atom::Variable(*a.min(b))) * Expression::Atom(Atom::Variable(*a.max(b))), *a.min(b) != *a),
+        (e1, e2) => (Expression::Multiply(Box::new(e1.clone()), Box::new(e2.clone())), false)
+    }
+}
+
+fn switch_atoms_ext_mul(map: &LinearMap<Atom, Expression, 8>) -> (Expression, bool) {
+    match (map.get(&Atom::Escape('*', 1)).unwrap(), map.get(&Atom::Escape('A', 1)).unwrap(), map.get(&Atom::Escape('A', 2)).unwrap(),) {
+        (e, Expression::Atom(Atom::Variable(a)), Expression::Atom(Atom::Numeric(b))) => (e.clone() * Expression::Atom(Atom::Numeric(*b)) * Expression::Atom(Atom::Variable(*a)), true),
+        (e, Expression::Atom(Atom::Variable(a)), Expression::Atom(Atom::Variable(b))) => (e.clone() * Expression::Atom(Atom::Variable(*a.min(b))) * Expression::Atom(Atom::Variable(*a.max(b))), *a.min(b) != *a),
+        (e1, e2, e3) => (e1.clone() * e2.clone() * e3.clone(), false),
+    }
+}
+
+fn denegate_internal(map: &LinearMap<Atom, Expression, 8>) -> (Expression, bool) {
+    match map.get(&Atom::Escape('A', 1)).unwrap() {
+        Expression::Atom(Atom::Numeric(n)) => match n {
+            Numeric::Integer(i) => if i < &0 {
+                (Expression::Negate(Box::new(Expression::Atom(Atom::Numeric(Numeric::Integer(-i))))), true)
+            } else {
+                (Expression::Atom(Atom::Numeric(Numeric::Integer(*i))), false)
+            },
+            Numeric::Decimal(f) => if f < &0.0 {
+                (Expression::Negate(Box::new(Expression::Atom(Atom::Numeric(Numeric::Decimal(-f))))), true)
+            } else {
+                (Expression::Atom(Atom::Numeric(Numeric::Decimal(*f))), false)
+            },
+            Numeric::Fraction(r1, r2) => if r1 < &0 && r2 < &0 {
+                (Expression::Atom(Atom::Numeric(Numeric::Fraction(-r1, -r2))), true)
+            } else if r1 < &0 {
+                (Expression::Negate(Box::new(Expression::Atom(Atom::Numeric(Numeric::Fraction(-r1, *r2))))), true)
+            } else if r2 < &0 {
+                (Expression::Negate(Box::new(Expression::Atom(Atom::Numeric(Numeric::Fraction(*r1, -r2))))), true)
+            } else {
+                (Expression::Atom(Atom::Numeric(Numeric::Fraction(*r1, *r2))), false)
+            },
+        },
+        e => (e.clone(), false),
+    }
+}
+
+//an AdaptableModifier that reduces identities in an expression tree
+pub fn reduce() -> AdaptableModifier {
+    AdaptableModifier::from_str_list(vec![
+        ("_*1 + _*1", "2 * _*1"),
+        ("_A1 * _*1 + _*1", "(1 + _A1) * _*1"),
+        ("_*1 + _A1 * _*1", "(1 + _A1) * _*1"),
```

Comment:
Just a point of thought does this really *reduce* the number of nodes in the tree and is it essential? The thing with simplification is that applying one rule earlier in the trace will cause a later simplification rule that should've been valid become invalid.

Reply:
oh, I fucked those ones up. it was not supposed to be like that.

Reply:
actually, nevermind. I think it is?

Reply:
yea they are correct, but do they really reduce the tree? since `a * b + b` -> `(a + 1) * b` checks out to be still the same tree depth

Reply:
(a+1)*b can be further reduced, while a*b+b cannot. for example 5*x+x -> (5+1)*x -> 6*x